Internal(Components): Bug with components lock mechanism [ED-21559]

MuliElementor opened the pull request:
## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

<!--start_gitstream_placeholder-->
### ✨ PR Description
Purpose: Fix component lock mechanism to prevent concurrent editing of the same component by multiple users.
Main changes:
- Renamed Lock_Component_Manager to Component_Lock_Manager for consistent naming
- Added validation to ensure operations only run on component post types
- Implemented get_updated_status method to correctly handle expired locks
- Added comprehensive unit tests for the lock manager functionality

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using.
💡 **Tip:** You can customize your AI Description using **Guidelines** [Learn how](https://docs.gitstream.cm/automation-actions/#describe-changes)</sub>
<!--end_gitstream_placeholder-->

MuliElementor committed:
fix inf loop
update class name

MuliElementor committed:
update

MuliElementor committed:
lint

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

MuliElementor committed:
Enhance component lock management by adding get_updated_status method to handle lock state updates. Update REST API to utilize the new method for checking lock status and user permissions. Clean up unused variables and improve unlock logic in Document_Lock_Manager.

MuliElementor committed:
lint

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

ronros-elementor commented on tests/phpunit/elementor/modules/components/test-component-lock-manager.php:
```diff
@@ -0,0 +1,340 @@
+		] );
+
+		// Create test documents of different types
+		$this->test_document_ids = $this->create_test_documents();
```
this naming is a bit strange, because it doesn't hold on document ids, one is of a document, the other is for component.

I'd suggest splitting the two, so that the create test document would just create a document, just like `create_test_component` only creates a component :)

```suggestion
		$this->test_document_id = $this->create_test_document();
		$this->test_component_id = $this->create_test_component();
```

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

MuliElementor committed:
add guard for components type
update tests

MuliElementor committed:
lint

elementorbot committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

MuliElementor committed:
Enhance component lock management by adding validation for component IDs and updating lock status checks. Introduce a new method to verify component type. Update tests to reflect changes in lock behavior for non-existent components.

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

MuliElementor committed:
Refactor: Remove unnecessary whitespace in components REST API methods for cleaner code.

MuliElementor committed:
Fix REST API to handle non-existent components gracefully

MuliElementor committed:
update

MuliElementor commented on modules/components/document-lock-manager.php:
```diff
@@ -86,13 +87,14 @@ public function lock( $document_id ) {
 
+		// Remove edit lock metadata
+		delete_post_meta( $document_id, self::LOCK_EDIT_LOCK_META );
+
```
yes

MuliElementor committed:
update

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

ronros-elementor commented on modules/components/component-lock-manager.php:
```diff
@@ -61,4 +65,38 @@ public function heartbeat_received( $response, $data ) {
+		}
+
+		return parent::lock( $document_id );
+	}
```
If it doesn't add any functionality that is relevant to components in particular - why do we need to override it? Even if theoretically it won't be an id of a component, but of a document - it would anyway go to the document-lock-manager (i.e. `parent::lock(...)`)

ronros-elementor commented on modules/components/components-rest-api.php:
```diff
@@ -243,12 +251,26 @@ private function get_lock_status( \WP_REST_Request $request ) {
+		}
+
+		return get_post_type( $post_id ) === Component_Document::TYPE;
+	}
```
Why do we need to validate if a component, if anyway we have this validation in the lock manager on every function call? I'd say - either remove it from there, or from here... Also I think that by only using get_post_type - we no longer need to verify the post even exists (but please make sure I'm right... but anyway in the lock manager's implementation - the check is only against the post type)

MuliElementor committed:
Refactor: Simplify lock status checks in Components REST API

- Removed redundant validation for component existence in get_lock_status and is_current_user_allow_to_edit methods.
- Cleaned up the code by eliminating the is_valid_component method, streamlining the logic for checking user permissions on component editing.
- Updated related unit tests to reflect the changes in lock status handling.

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix

MuliElementor committed:
Enhancement: Add error handling for component lock and unlock operations in Components REST API

- Implemented try-catch blocks in lock_component and unlock_component methods to handle exceptions and log errors.
- Updated get_lock_status and is_current_user_allow_to_edit methods to include error handling for lock status retrieval.
- Improved user feedback by returning appropriate error messages when locking or unlocking components fails.

MuliElementor committed:
Merge branch 'main' into ED-21063-prevent-multiple-users-editing-the-same-component-at-the-same-time-fix
